**Practice Linear Programming**

Feel free to refer to the demo done in lecture: https://colab.research.google.com/drive/10yYw97pNP6SoNrXjHcijKnDB7oKyLkv5?usp=sharing



Suppose Cornell dinning halls produce **80 kg of food waste per day**.

They can either:

1. **Compost** it, with a cost of $0.12/kg and the impact of 2 points per kg (environmental impact :)

2. **Donate** it to a local food bank with the cost of $0.25/kg and impact of 5 points per kg (social impact :)

However Cornell dinning halls must 1. process **all 80 kg** of food waste every day 2. stay within a **$15 budget**

Goal: Use Linear Programming to decide how much food to compost and how much to donate in order to **maximize total impact**.

Please make sure you answer the following:
  - How many kg of food to **compost**
  - How many kg of food to **donate**
  - What is the **total impact score**


In [1]:
pip install scipy

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 37.6/37.6 MB 52.6 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.4/16.4 MB 55.8 MB/s eta 0:00:00:00:01
Note: you may need to restart the kernel to use updated packages.


In [2]:

from scipy.optimize import linprog

In [3]:
# The objective is to maximize social impact. Compost social impact per kg = 2; Donate social impact per kg= 5
objective = [-2, -5]

In [7]:
# Constraints:
# $.12/kg * compost + $.25/kg * donate ≤ 15 (budget limit)

constraint_coeffs = [
    [.12, .25],  # Cost coefficients
]

constraint_limits = [15]  # constraints, given on the RHS of the equations

In [8]:
# Equality constraints, so 1x + 1y = 80
a_eq = [[1, 1]]
b_eq = [80]

In [9]:
# number of food to compost or donate must both be non-negative

x_bounds = (0, None)
y_bounds = (0, None)

In [11]:
# Solve the problem
solution = linprog(objective, A_ub= constraint_coeffs, b_ub= constraint_limits, A_eq= a_eq, b_eq=b_eq,
                   bounds=[x_bounds, y_bounds], method='highs')


In [14]:
# Show results
if solution.success:
    print("Optimal solution found:")
    #index 0 for solar panel, first item in the solution list
    print(f"# of kg to compost: {solution.x[0]}")
    print(f"# of kg to donate: {solution.x[1]}")
    print(f"Total Impact Score: {-solution.fun}")
else:
    print("No solution found.") #when .success == False

Optimal solution found:
# of kg to compost: 38.46153846153846
# of kg to donate: 41.53846153846154
Total Impact Score: 284.61538461538464


So, 
number of kg to compost: 38.46153846153846
number of kg to donate: 41.53846153846154
Total Impact Score: 284.61538461538464

Reflection (Required)

**How can linear programming be incorporated into projects you are currently working on or might work on in the future in GreenClub?**  

My response: Linear programming can be really helpful in the Composting TECH NME project I'm working on in GreenClub. It can help us figure out the most efficient ways to compost by automating the process, which is much faster than doing the calculations by hand. In the future, I’d love to use linear programming to improve sustainability in business. It could be a great tool for optimizing food waste management or finding the most eco-friendly methods for production. It’s a really powerful way to make sure we’re making the best decisions when it comes to sustainability.

